Mathlib/Order/SuccPred/Tree.lean

```diff
@@ -0,0 +1,226 @@
+-/
+def SubRootedTree (t : RootedTree) : Type* := t
+
+/--
+The root of a `SubRootedTree`.
+-/
+def SubRootedTree.root {t : RootedTree} (v : SubRootedTree t) : t := v
+
+/--
+The `SubRootedTree` rooted at a given node.
+-/
+def RootedTree.subtree (t : RootedTree) (r : t) : SubRootedTree t := r
+
+@[simp]
+lemma RootedTree.root_subtree (t : RootedTree) (r : t) : (t.subtree r).root = r := rfl
+
+@[simp]
+lemma RootedTree.subtree_root (t : RootedTree) (v : SubRootedTree t) : t.subtree v.root = v := rfl
+
+@[ext]
+lemma SubRootedTree.ext {t : RootedTree} (v₁ v₂ : SubRootedTree t)
+    (h : v₁.root = v₂.root) : v₁ = v₂ := h
+
+instance (t : RootedTree) : SetLike (SubRootedTree t) t where
+  coe v := Set.Ici v.root
+  coe_injective' a₁ a₂ h := by
+    simpa only [Set.Ici_inj, ← SubRootedTree.ext_iff] using h
+
+lemma mem_iff {t : RootedTree} {r : SubRootedTree t} {v : t} :
+    v ∈ r ↔ r.root ≤ v := Iff.rfl
+
+/--
+The coercion from a `SubRootedTree` to a `RootedTree`.
+-/
+@[coe, reducible]
+noncomputable def coeTree {t : RootedTree} (r : SubRootedTree t) : RootedTree where
+  α := Set.Ici r.root
+
+noncomputable instance (t : RootedTree) : CoeOut (SubRootedTree t) RootedTree :=
+  ⟨coeTree⟩
+
+@[simp]
+lemma bot_mem_iff {t : RootedTree} (r : SubRootedTree t) :
+    ⊥ ∈ r ↔ r.root = ⊥ := by
+  simp [mem_iff]
+
+/--
+All of the immediate subtrees of a given rooted tree.
+-/
+def RootedTree.subtrees (t : RootedTree) : Set (SubRootedTree t) :=
+  {x | IsAtom x.root}
+
+variable {t : RootedTree}
+
+lemma root_ne_bot_of_mem_subtrees (r : SubRootedTree t) (hr : r ∈ t.subtrees) :
+    r.root ≠ ⊥ := by
+  simp only [RootedTree.subtrees, Set.mem_setOf_eq] at hr
+  exact hr.1
+
+lemma subtrees_inf_eq_bot_iff {t₁ t₂ : SubRootedTree t}
+    (ht₁ : t₁ ∈ t.subtrees) (ht₂ : t₂ ∈ t.subtrees) (v₁ v₂ : t) (h₁ : v₁ ∈ t₁)
+    (h₂ : v₂ ∈ t₂) :
+    v₁ ⊓ v₂ = ⊥ ↔ t₁ ≠ t₂ where
+  mp h := by
+    intro nh
+    have : t₁.root ≤ (v₁ : t) ⊓ (v₂ : t) := by
+      simp only [le_inf_iff]
+      exact ⟨h₁, nh ▸ h₂⟩
+    rw [h] at this
+    simp only [le_bot_iff] at this
+    exact root_ne_bot_of_mem_subtrees t₁ ht₁ this
+  mpr h := by
+    rw [mem_iff] at h₁ h₂
+    contrapose! h
+    rw [← bot_lt_iff_ne_bot] at h
+    rcases lt_or_le_of_directed (by simp : v₁ ⊓ v₂ ≤ v₁) h₁ with oh | oh
+    · simp_all [RootedTree.subtrees, IsAtom.lt_iff]
+    rw [le_inf_iff] at oh
+    ext
+    simpa only [ht₂.le_iff_eq ht₁.1, ht₁.le_iff_eq ht₂.1, eq_comm, or_self] using
+      le_total_of_directed oh.2 h₂
+
+lemma subtrees_disjoint : t.subtrees.PairwiseDisjoint ((↑) : _ → Set t) := by
+  intro t₁ ht₁ t₂ ht₂ h
+  rw [Function.onFun_apply, Set.disjoint_left]
+  intro a ha hb
+  rw [← subtrees_inf_eq_bot_iff ht₁ ht₂ a a ha hb] at h
+  simp only [le_refl, inf_of_le_left] at h
+  subst h
+  simp only [SetLike.mem_coe, bot_mem_iff] at ha
+  exact root_ne_bot_of_mem_subtrees t₁ ht₁ ha
+
+/--
+The subtree of `t` containing `r`, or all of `t` if `r` is the root.
```

Comment:
```suggestion
The immediate subtree of `t` containing `r`, or all of `t` if `r` is the root.
```
Else that would be the subtree rooted at `r`, right?

I would also suggest renaming `r` to something else, since it's not the root of any tree under consideration